### ISSUE TITLE
[ENH] Feature importance/explainability for *RocketMultivariate()-based classifiers

### ISSUE THREAD

**Issue by varelasaraiva** [sktime/sktime#4336]
It's common practice to do feature analysis when building tabular classification models using methods like SHAP, to debug modelling issues, help explain model outputs to business stakeholders, etc.

The ROCKET family of time series feature extraction algorithms has proven most useful to perform multivariate time series classification but the 3-dimensional input datasets are not recognised by `shap`.

One workaround is to do univariate analysis, whereby we iterate over all features, convert the matrix of shape (instances, timepoints) to a tabular dataset and get a matrix of SHAP values for (instances, timepoints). If we calculate mean absolute SHAP along both axes, that can potentially be a measurement of feature importance. Moreover, if we calculate mean absolute SHAP for each timepoint, we could also discuss e.g. if some hours of the day are more decisive than others. The downside is that we can't do any multivariate analysis like feature interactions, etc. Not to mention that `shap` is completely blind to the properties of a time series (autocorrelation, etc).

I have found this paper (https://arxiv.org/abs/2210.02176) describing the Time consistent SHAP but there is no public Python implementation. I have also been looking at TimeSHAP (https://github.com/feedzai/timeshap) but it only seems to support pytorch RNNs. Both these methods are perhaps overkill for TSC.

Is there any recommended method to calculate feature importance for multivariate timeseries classification using `*RocketMultivariate()`?

**Comment by fkiraly** [comment]
FYI @angus924 (author) - pointers appreciated!

**Comment by fkiraly** [comment]
I found this package https://github.com/mlgig/mrsqm by @lnthach and @heerme - this is a different transformer, but has a `get_saliency_map` method which is a temporal feature importance map.

This is only compliant with the "old" `sktime` classifier interface, so I wonder whether there are plans to update it?

I suppose the way to do it would be the same as for MrSEQL as discussed in this issue:
https://github.com/sktime/sktime/issues/4296

**Comment by varelasaraiva** [comment]
@kwulffert and I have been using the following workaround for this issue. SHAP seems to induce perturbations element-wise so in theory there seems to be no reason this couldn’t work for tensors of rank > 2. 

The main idea is to reshape the initial time series tensor as a 2D matrix, pass it to SHAP and use a sklearn pipeline as the function that returns model outputs. Within this pipeline, the 2D matrix is reshaped to the original tensor, converted to tabular with MiniROCKET and fed into the model which returns tabular outputs. 

SHAP associates the original 2D matrix with the tabular outputs and returns a 2D matrix of SHAP values with the shape of the 2D matrix. If we then reshape the 2D SHAP matrix to a 3D SHAP tensor of the initial shape, this tensor contains SHAP values for each observation, timepoint and feature. 

This allows aggregating along the feature axis and use SHAP values as a measure of feature importance, but also along the time axis and use SHAP to determine e.g. which hours of the day contribute more to the model output.

Example code can found in https://gist.github.com/varelasaraiva/fd16a3108a5bb1ae47b27443ca20933f.

**Comment by fkiraly** [comment]
hm, if you are working with `shap`, what should also work is just directly doing sth like

```python
explainer = shap.KernelExplainer(my_clf.predict_proba, X_train)
shap_values = explainer.shap_values(X_test)
```

if you provide `X_train`, `X_test` in 2D numpy format.
This wouldn't work currently with `pandas` though, since a `pandas` `DataFrame` is always interpreted as single time series.

That's quite similar to what happens in your gist already.

I'm wondering whether it would be nice:

* to have the code in your `gist`, in slightly more abstract design (e.g., taking arbitrary classifier, also returning `shap_values` return) in `sktime` as a utility in an `explain` module or similar
* to have sth similar, but as an `sktime` transformer. That would take a classifier as a parameter, and in `transform` produce the SHAP tensor, in exactly the same format as the input

What do you think, @varelasaraiva?

**Comment by pascalamadeus** [comment]
Hey @fkiraly and @varelasaraiva,

are there any new implementations making SHAP work for sktime ROCKET classifier on multivariate time series? 

Is the workaround with 2D numpy format still the suggested implementation?

Thanks you!

**Comment by fkiraly** [comment]
I think this has not been implemented so far, @pascalamadeus - want to give it a try, based on @varelasaraiva's gist?

**Comment by pascalamadeus** [comment]
@fkiraly, I have applied the approach given in @varelasaraiva´s gist to my example problem and it works in combination with the ROCKET transformer as workaround. Thanks @varelasaraiva for the code! However, I am experiencing the following fundamental problem: the shap values are very low (maximum shap value of around 0.17 when using instances of 15 timepoints each). After an intensive research in the literature, I found the following problem: "numerical underflow". Citing Villani et al. [https://arxiv.org/abs/2210.02176](url) :

> "Secondly, time series models often take windows of data; in the presence of **long input windows the computation of KernelSHAP becomes impossible.** As we will see later, this is due to **numerical underflow in the normalisation coefficient present in the calculations required for KernelSHAP**. Indeed, in practice, it is common to find very long lookback windows."
> 
> [...]
> 
> "In the context of feed-forward neural network, where the input features exist in distinct spaces, features can reasonably be considered as different players in a cooperative game. However, **in the time series domain, the same reasoning will lead SHAP to consider the same feature at different time steps as different players in a game**."
> 

I could confirm this issue in my experiments: I´ve tried different lengths of instances (minimum was limited to 9 timepoints due to the ROCKET implementation in sktime). I found that shorter instances lead to higher shap values and longer instances to lower shap values (close to zero). It seems to be a specific problem with the kernelExplainer which I think has to be fixed before implementaion in sktime. 

Villani et al. suggest "Time Consistent SHAP Values" to overcome this issue. However, I haven´t found any implementation of this approach yet. Would be interested in your thoughts on that!
